### ISSUE TITLE
Should requests be queued?

### ISSUE THREAD

**Issue by zenparsing** [zenparsing/async-iteration#14]
Currently, we queue calls to `next`, `throw`, or `return` if the generator is currently paused on an await (or running).  You can do things like this:

```js
// Get an async iterator which represents a write stream to a file
let sink = createFileWriteGenerator(path);

sink.next(buffer1);  // Write a chunk without waiting
sink.next(buffer2);  // Write another chunk without waiting
await sink.return(); // Close the file and wait for the handle to close
```

But there are a couple of issues here:

1. Unbounded queues are kinda bad.
1. It's somewhat incongruent with the way that regular generators work.  They throw if the generator is currently executing.
1. It can, in principle, lead to some weird infinite waiting cases.

```js
let iter = g();
iter.next(); // Never resolves
async function *g() {
    await iter.next(); // Never resolves
}
```

The other option is to return a rejected promise if the generator is currently running or paused on an await expression.

Thoughts?

@jhusain
@domenic
@bterlson
@littledan

**Comment by domenic** [comment]
The current design makes the most sense to me. Unbounded queues are the name of the game here and aren't really a bad thing in general. (In many senses an array is an unbounded queue in JavaScript.)

Regular generators aren't async, so the idea of doing something while the generator is currently executing is not a normal occurence. Or maybe I am misunderstanding this point.

Infinite waiting cases seem easy to get with circular promise situations in general, and this is no difference.

**Comment by eggers** [comment]
The case you gave would be an error because aren't functions are lifted to the top of the block,  so that would actually be:

```js
async function *g() {
    await iter.next(); // throws
}

let iter = g();
iter.next(); // resolves with error
```

**Comment by zenparsing** [comment]
Out of curiosity, I updated [esdown](https://github.com/zenparsing/esdown) to throw if there is an in-flight request and then analyzed the impact to my previous work using async iterators.  In particular, I have a zip/tar utility named [ziptar](https://github.com/zenparsing/ziptar) which uses async iterators quite heavily.

After disabling queuing, ziptar definitely broke.  I was taking advantage of queuing in the following two cases:

1. In order to "prime" an async generator acting as a data sink, I was calling an initial `next()` without awaiting the result.
1. There was a situation where I needed to write a couple of buffers to a stream and then close it.  I was only awaiting the final `return` and not the preceding `next` calls.

Once I fixed these issues, I saw no further problems with ziptar.

This leads me to believe that queuing is convenient and will definitely be taken advantage of, but is probably not essential to the design.  I don't think it represents a footgun, since you have to intentionally "fill" the queue.

**Comment by RangerMauve** [comment]
Personally, I like the idea of queueing as long as all the promises get rejected if the async iterator breaks prematurely. Though I can see the scaryness in unbound queues, I've found it to be a useful mechanic in [my own code](https://github.com/RangerMauve/promise-channel)

**Comment by bmeck** [comment]
@zenparsing as long as the Promise returned is a builtin ES2015 Promise I think it will be fine. If you allowed Promise subclasses to be returned via queueing then you could introduce side effects.

I would expect a cancellation signal (ala .finally without rejection more than seeing all unbound promises getting rejected).

**Comment by hayes** [comment]
something that might be worth considering is parallel consumption of an async iterator. I'm not sure if its desired, but would have a big impact on this question.

I'm not sure if my interpretation is exactly right, but this example should at least illustrate my question
```javascript
async function *count(n) {
  for (let i = 1; i <= n; ++i) {
    yield Promise.resolve(i)
  }
}

async function sum (itr) {
  let total = 0
  for await (let val of itr) {
    total += val
  }

  return total
}

let counter = count(5);

// With queueing
sum(counter).then(val => console.log(val) // logs 9 (1 + 3 + 5)
sum(counter).then(val => console.log(val) // logs 6 ( 2 + 4)
// Without queueing
sum(counter).then(val => console.log(val) // logs 15 ( 1 + 2 + 3 + 4 + 5)
sum(counter).catch(val => console.log(val) // some error about already having queued a call to next?
```

**Comment by hayes** [comment]
Personally parallel consumption with queuing seems like it would be a huge footgun, and I don't think there would be an easy way to differentiate 2 separate consumers from a single complex consumer, which causes me to lean strongly towards not queuing calls. One part I am not entirely clear on is how queueing interacts with return/throw calls. I am guessing when queued they would be queued in the same queue as next calls, but if you go with the not queueing route do those calls throw if there is a pending next?

**Comment by zenparsing** [comment]
@hayes Thanks for the code example.  I'll need to think about this for a bit.

> I am guessing when queued they would be queued in the same queue as next calls

Yes.

> but if you go with the not queueing route do those calls throw if there is a pending next?

I would think so, yes.

**Comment by eggers** [comment]
@hayes I would say it should log 15 twice. IMO, each new `for await...of` should create a new subscription to the generator (which would be discarded at the end of the loop.) That's not how sync `for..of` loops are currently treated, but it would be more inline with the es7 observable proposal, and specifically discussion here: zenparsing/es-observable#73

But then again, IMO `async *()` should return an observable, not a generator since they both serve the same basic purpose and the Observable proposal is further along than this one.

**Comment by hayes** [comment]
@eggers I think the solution there would be some type of helper to convert from an async iterator to an Observable.
The thing thats great about async iterators is that it uses a pull model. Observables don't really address the lazy use cases I want async iterators for.  There is a lot of overlap between usecases for async Iterators and Observables,  but fundamentally they are not the same, and can exist in parallel without hard dependencies on each other. If they both are approved, utility function (like Observalbe.from) to easily combine them would be awesome).

**Comment by zenparsing** [comment]
@eggers We definitely want async and sync for-of loops to be as similar as possible.  Other than that, I agree with what @hayes said.  Observables could potentially be convertable to async-iterators via a `Symbol.asyncIterator` method or specific combinators, and async iterators could be convertable to observables via Observable.from.

**Comment by zenparsing** [comment]
@hayes How is the footgun you pointed out any different from this footgun?

```js
// Regular generator function definition
function *count(n) {
  for (let i = 1; i <= n; ++i) {
    yield i
  }
}

async function sum (itr) {
  let total = 0
  for (let val of itr) { // <== Sync for-of
    await 1; // <== Note this
    total += val
  }

  return total
}

let counter = count(5);

sum(counter).then(val => console.log(val))
sum(counter).then(val => console.log(val))
```

In other words, is this an issue because of queuing, or an issue because we've shared an iterator across two async consumers?

**Comment by hayes** [comment]
that's an interesting point, although without queuing I believe this would still error. 

sum would start to execute for the first invocation
count would be queued to resume
sum would start its second iteration
a second call to next would attempt to queue another iteration of count before the previous count iteration was invoked.

This is based on an assumption I have not verified, that async generator functions are resumed asynchronously

**Comment by hayes** [comment]
I do see your general point though, but I don't really have any suggestions about how to prevent sharing an async iterator across multiple consumers.

**Comment by hayes** [comment]
Before I found this proposal today, my idea of how something like this would be built was something like

```
async function * () {} returns  Promise<AsyncItoration>

interface AsyncItoration {
  done: Boolean,
  value: any,
  next: Function<Promise<AsyncItoration>>?
}
```

the idea for consumption was similar, but syntax was probably worse.

would be cool because you don't have to worry about queueing since you can only get the next value from the previous. You can easily have multiple consumers, and promises will cache/store your values for you for future consumers.

The downsides are that any reference holds on to all future items in the series which makes implementing a for of that doesn't hold onto everything really hard (definitely a deal breaker). No return/throw equivalent.

**Comment by hayes** [comment]
My point with the previous response being, it would be good if the api prevented consumers from getting promises for more than one item at a time.  2 ideas come to mind, neither of which seemed great to me.

either `next()` could return the same promise for the next item multiple times (until next is resolved) or next itself could be the promise for the next thing at any given time.  (having a mutable object seems like a bad idea).

This would allow multiple consumers, and even consumers to be added part way through an iteration without any issues.

**Comment by zenparsing** [comment]
> either next() could return the same promise for the next item multiple times (until next is resolved) 

I don't think that would work.  What if I supply a different argument to the `next` call?

> next itself could be the promise for the next thing at any given time

We don't want to start work before the consumer has requested it with a call to `next` though.  And again, consider that `next` can accept an argument.

> This would allow multiple consumers, and even consumers to be added part way through an iteration without any issues.

Take a look at that code I posted upthread.  It doesn't have any async generators/iterators in it at all, at yet has the same issues that you point out.  The situation arises, not from the semantics of async iterators, but from the fact that we have shared a stateful object among two asynchronous (and possibly racy) operations.  

It seems to me that perhaps queueing provides the most sensible semantics in such a case, because otherwise one of the consumers would end up getting an error in a non-deterministic fashion.  (The ordering of the calls to `next` might change depending upon variances in how long each async operation takes in between `next` calls.)
